src/api/valence-api.js

```diff
@@ -0,0 +1,392 @@
+		console.log(`Updating topic file: '${fileName}'`);
+
+		const formData = new FormData();
+		formData.append(
+			'file',
+			data,
+			{ filename: `${fileName}` }
+		);
+
+		if (this._isDryRun) {
+			return topic;
+		}
+
+		const response = await this._fetch(
+			signedUrl,
+			{
+				method: 'PUT',
+				headers: {
+					'Content-Type': `multipart/form-data; boundary=${formData.getBoundary()}`
+				},
+				body: formData
+			});
+
+		this._assertResponse(response);
+
+		return this._convertTopic(topic, self.Id);
+	}
+
+	async assertQuiz(instanceUrl, orgUnit, quiz, module) {
+		const content = await this._getContent(instanceUrl, orgUnit, module);
+		const self = Array.isArray(content) && content.find(x => x.Type === 1 && x.TopicType === 3 && x.Title === quiz.title);
+
+		if (self) {
+			// Nothing to do, the quicklink exists
+			return quiz;
+		}
+
+		const quizzes = await this._getQuizzes(instanceUrl, orgUnit);
+		const quizItem = quizzes.Objects && Array.isArray(quizzes.Objects) && quizzes.Objects.find(x => x.Name === quiz.title);
+
+		return this._createQuizTopic(instanceUrl, orgUnit, { module, quiz, quizItem });
+	}
+
+	async _createQuizTopic(instanceUrl, orgUnit, { module, quiz, quizItem }) {
+		const url = new URL(`/d2l/api/le/${LEVersion}/${orgUnit.Identifier}/content/modules/${module.Id || module.id}/structure/`, instanceUrl);
+		const signedUrl = this._valence.createAuthenticatedUrl(url, 'POST');
+
+		console.log(`Creating quiz topic: '${quiz.title}'`);
+
+		const rcode = quizItem.ActivityId.split('/').slice(-1)[0];
+		const topic = ContentFactory.createTopic({
+			title: quiz.title,
+			topicType: 3,
+			url: `/d2l/common/dialogs/quickLink/quickLink.d2l?ou=${orgUnit.Identifier}&type=quiz&rcode=${rcode}`,
+			isExempt: !quiz.isRequired
+		});
+
+		if (this._dryRun) {
+			return {
+				...quiz,
+				id: DryRunId
+			};
+		}
+
+		const response = await this._fetch(
+			signedUrl,
+			{
+				method: 'POST',
+				headers: {
+					'Content-Type': 'application/json'
+				},
+				body: JSON.stringify(topic)
+			}
+		);
+
+		this._assertResponse(response);
+
+		const json = await response.json();
+
+		return this._convertTopic(quiz, json.Id);
+	}
+
+	async getOrgUnit(instanceUrl, orgUnitId) {
+		const url = new URL(`/d2l/api/lp/${LPVersion}/courses/${orgUnitId}`, instanceUrl);
+		const signedUrl = this._valence.createAuthenticatedUrl(url, 'GET');
+
+		const response = await this._fetch(signedUrl);
+
+		this._assertResponse(response);
+
+		return response.json();
+	}
+
+	async whoAmI(instanceUrl) {
+		const url = new URL(`/d2l/api/lp/${LPVersion}/users/whoami`, instanceUrl);
+		const signedUrl = this._valence.createAuthenticatedUrl(url, 'GET');
+
+		const response = await this._fetch(signedUrl);
+
+		this._assertResponse(response);
+
+		return response.json();
+	}
+
+	_assertResponse(response) {
+		if (response.ok) {
+			return;
+		}
+
+		throw new Error(response.statusText);
+	}
+
+	_convertModule(module, id) {
```

Comment:
This is shaped this way to drop the `children` property as the data is passed around.